designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,213 @@
+| Authors       | Creation Date | Status      | Extra |
+|---------------|---------------|-------------|---|
+| @name | date | Implementeble | - |
+
+# Auto-generate Helm Chart Plugin
+===================
+
+The proposal is to add a new alpha command to auto-generate a helm chart for a given kustomize directory created by kubebuilder.
+
+## Example
+
+By running the following command, the plugin will generate a helm chart from the specific kustomize directory and output it to the directory specified by the `--output` flag.
+
+```shell
+kubebuilder alpha generate-helm-chart --from=<path> --output=<path>
+```
+
+## Open Questions [optional]
+
+TBF
+
+## Summary
+
+TBF
+
+## Motivation
+
+Currently, kubebuilder only supports the distribution of the operator by the kustomize. However, many well-known operators contains the distribution of helm charts:
+
+- [mongodb](https://artifacthub.io/packages/helm/mongodb-helm-charts/community-operator
+)
+- [cert-manager](https://cert-manager.io/v1.6-docs/installation/helm/#1-add-the-jetstack-helm-repository)
+- [prometheus](https://bitnami.com/stack/prometheus-operator/helm)
+
+The proposed command will auto-generate a helm chart from a kustomize directory created by kubebuilder with minimizing the manual effort. The generated helm chart will cover the frequently-used configurations.
+
+Context:
+
+- See the discussion [Is there a good solution/tool to generate a new helm chart from the manifests?](https://github.com/kubernetes-sigs/kubebuilder/discussions/3074)
+
+### Goals
+
+- Help developers to generate a helm chart from a kustomize directory.
+- Generate a helm chart without human interactions once the kustomize directory updated.
+- Reserve the ability to customize the helm chart by the developer.
+
+### Non-Goals
+
+- Generate the helm chart from a kustomize directory not created by kubebuilder.
+- Generate the fine-grained configurations.
+
+## Proposal
+
+To reduce the user adaptation, the command
+leverage the existing kustomize default directory and create the similar layout of the helm chart from it. The generated `values.yaml` is as follows:
+
+```yaml
+# global configurations
+nameOverride: ""
+
+fullnameOverride: ""
+
+crd:
+  create: true
```

Comment:
This one for example should be scaffold as false when we run the init command with the plugin or the edit and no CRDs were created yet. See that we track all in the PROJECT file so we must to check in order to properly add false or true.

Then, when we run kubebuilder create api we should ensure that it has true if a resource was generated.

Reply:
I've seen a couple of controllers offering the option not to install CRDs via the chart, but honestly never seen an use case for it being disabled. @dashanji have you ever come across the need to disable it?

If it really is uncommon, I would push for not exposing this field in the first place, and letting the small percentage of users that need this, add it manually.

For example, we don't conditionally install CRDs via the generated kustomize: they're always installed. I think it's a good idea to follow the same decisions that were taken for Kustomize here in Helm, so that users are not surprised with inconsistent behavior.

Reply:
> I've seen a couple of controllers offering the option not to install CRDs via the chart, but honestly never seen an use case for it being disabled. @dashanji have you ever come across the need to disable it?

Then that is a too-specific edge case I think
But in the future if needed we can revisit.

Reply:
>For example, we don't conditionally install CRDs via the generated kustomize: they're always installed

How they are installed already?
If you create the API/CRD in the Operator how it can exist on the cluster in a common scenario before it be installed?

Reply:
> > For example, we don't conditionally install CRDs via the generated kustomize: they're always installed
> 
> How they are installed already? If you create the API/CRD in the Operator how it can exist on the cluster in a common scenario before it be installed?

I mean to say that there is no way to conditionally install the CRD via Kustomize, as an argument that there shouldn't be one for Helm either. I have never came across a scenario on which the CRD was installed prior to the controller being deployed.

Reply:
If you did not run `make install` to install the CRDs when you deploy the Manager/Operator (which has the controllers) `make deploy` then the manager will fail in the initialization.

Reply:
> If you did not run `make install` to install the CRDs when you deploy the Manager/Operator (which has the controllers) `make deploy` then the manager will fail in the initialization.

True, fair enough, that's not a good argument I gave. I still think we shouldn't add this `crd.create` in the values file, because it seems to be only there to facilitate an uncommon use-case for a Helm Chart.

Reply:
So, same controller installed multiple times in the same cluster? Don't they conflict with each other, in the sense that they all reconcile the same resources? Just trying to understand. In our case, we only install each controller once per cluster, so I guess that's why we never came across this.

Reply:
> So, same controller installed multiple times in the same cluster? Don't they conflict with each other, in the sense that they all reconcile the same resources?

Yep, especially when we use cert-manager as a sub chart, the cert-manager component will exist under each namespace along with the operator. Also, there is a situation where you do not use make `undeploy/uninstall` to delete resources such as CRD and webhook. Instead, you directly use a command like `kubectl delete deployment xxx` to delete the controller.

Reply:
Got it, I can see how this is useful for controllers that will be used as a dependency. 

We took a different approach, in which we just assume the CRD from another controller we require exists (and if it doesn't, it should be fixed elsewhere). For example, in controllers that depend on cert-manager, we just add our Issuers, Certificates, etc, but no CRD. The CRD should be provided by the cert-manager installation. This way we don't need to always define all CRDs we depend on, only the ones we provide in each specific controller.

But the bottom line is: should kubebuilder adopt this, given that there are many different ways to solve this particular problem? Since many charts provide this, I'd be ok with it being here as well and no longer have a strong opinion on this.

Reply:
I think if we have cert-manager enable, then we should scaffold it as a dependency in helm-charts.

Reply:
How to deal with CertManager and Prometheus?

^ I think we need add those in the OpenQuestions